jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java

```diff
@@ -81,76 +86,181 @@ public Object getVariable(String name) {
+                    variable,
+                    new Handler<>(
+                            getProcessEventSupport(),
+                            getProcessInstance(),
+                            name,
+                            prefixed(name),
+                            instancePrefixed(name)
+                    ));
+        }
+    }
+
+    protected ProcessEventSupport getProcessEventSupport() {
+        return ((InternalProcessRuntime) getProcessInstance()
+                .getKnowledgeRuntime()
+                .getProcessRuntime())
+                .getProcessEventSupport();
+    }
+
+    private static class Handler<T> implements ReferenceVariableInstance.OnSetHandler<T> {
+
+        transient final ProcessEventSupport processEventSupport;
+        transient final ProcessInstance processInstance;
+        final String name;
+        private final String prefixed;
+        private final String instancePrefixed;
+
+        public Handler(ProcessEventSupport processEventSupport, ProcessInstance processInstance, String name, String prefixed, String instancePrefixed) {
+            this.processEventSupport = processEventSupport;
+            this.processInstance = processInstance;
+            this.name = name;
+            this.prefixed = prefixed;
+            this.instancePrefixed = instancePrefixed;
+        }
+
+        @Override
+        public void before(Object oldValue, Object newValue) {
+            processEventSupport.fireBeforeVariableChanged(
+                    prefixed,
+                    instancePrefixed,
+                    oldValue,
+                    newValue,
+                    processInstance,
+                    processInstance.getKnowledgeRuntime());
+        }
+
+        @Override
+        public void after(Object oldValue, Object newValue) {
+            processEventSupport.fireAfterVariableChanged(
+                    prefixed,
+                    instancePrefixed,
+                    oldValue,
+                    newValue,
+                    processInstance,
+                    processInstance.getKnowledgeRuntime());
+        }
+    }
+
+    /**
+     * Assign a value reference to a declared variable instance.
+     * @throws IllegalArgumentException if the given variableName has not been declared in the process
+     */
+    public VariableInstance<?> assignVariableInstance(String variableName, ValueReference reference) {
+        VariableScope variableScope = getVariableScope();
+        Variable variable = variableScope.findVariable(variableName);
+        if (variable == null) {
+            variable = new Variable(variableName);
+            variable.setMetaData("on-the-fly", true);
+        }
+        variableScope.validateVariable(
+                getProcessInstance().getProcessName(),
+                variableName,
+                reference.get());
+        VariableInstance<?> instance = getVariableInstance(variableName);
+        instance.setReference(reference);
```

Comment:
yes exactly. This is all done internally; the API requiring a ProcessVariables instance instead of a Map<String,Object> is not not published to kie-api for now